crates/circuit/src/circuit_data.rs

```diff
@@ -459,21 +459,14 @@ impl CircuitData {
         clbits: Option<&Bound<PyAny>>,
     ) -> PyResult<()> {
         let mut temp = CircuitData::new(py, qubits, clbits, None, 0)?;
-        if temp.qubits_native.len() < self.qubits_native.len() {
-            return Err(PyValueError::new_err(format!(
-                "Replacement 'qubits' of size {:?} must contain at least {:?} bits.",
-                temp.qubits_native.len(),
-                self.qubits_native.len(),
-            )));
-        }
-        if temp.clbits_native.len() < self.clbits_native.len() {
-            return Err(PyValueError::new_err(format!(
-                "Replacement 'clbits' of size {:?} must contain at least {:?} bits.",
-                temp.clbits_native.len(),
-                self.clbits_native.len(),
-            )));
-        }
         if qubits.is_some() {
+            if temp.qubits_native.len() < self.qubits_native.len() {
+                return Err(PyValueError::new_err(format!(
+                    "Replacement 'qubits' of size {:?} must contain at least {:?} bits.",
+                    temp.qubits_native.len(),
+                    self.qubits_native.len(),
+                )));
+            }
```

Comment:
Was it coming up somewhere that we were trying to do this check when qubits was none? I would assume `qubits_native.len()` would just eval to 0 then and we'd only trigger the error if there was a mismatch. It doesn't really matter I guess though since this is also obviously correct too.

Reply:
There was some point at which we were replacing either qubits or clbits but not both - I don't remember which, but it was causing a hard failure because the previous logic was wrong in its checks in that case.